pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineActionTest.java

```diff
@@ -0,0 +1,649 @@
+        ChangeLogSet.Entry failingEntry = failingChanges.iterator().next();
+        assertNotNull(failingEntry);
+
+        // Make another commit
+        sampleRepo.write("otherNewFile", "exists");
+        sampleRepo.git("add", "otherNewFile");
+        sampleRepo.git("commit", "--message=poof");
+
+        HtmlPage redirect = restartFromStageInUI(failing, "restart");
+
+        assertNotNull(redirect);
+        assertEquals(p.getAbsoluteUrl(), redirect.getUrl().toString());
+
+        j.waitUntilNoActivity();
+        WorkflowRun b3 = p.getBuildByNumber(3);
+        assertNotNull(b3);
+        j.assertBuildStatusSuccess(b3);
+        j.assertLogContains("Odd numbered build, success", b3);
+        j.assertLogContains("Stage \"skip-on-restart\" skipped due to this build restarting at stage \"restart\"", b3);
+        j.assertLogContains("Now we're post-restart", b3);
+        j.assertLogNotContains("This shouldn't show up on third run", b3);
+
+        assertNotNull(b3.getChangeSets());
+        assertFalse(b3.getChangeSets().isEmpty());
+        assertEquals(1, b3.getChangeSets().size());
+        ChangeLogSet<? extends ChangeLogSet.Entry> b3Changes = failing.getChangeSets().get(0);
+        assertEquals(1, b3Changes.getItems().length);
+        ChangeLogSet.Entry b3Entry = b3Changes.iterator().next();
+        assertNotNull(b3Entry);
+
+        assertEquals(failingEntry.getCommitId(), b3Entry.getCommitId());
+    }
+
+    @Issue("JENKINS-45455")
+    @Test
+    public void isRestartedRunCondition() throws Exception {
+        WorkflowRun original = expect("restart", "isRestartedRunCondition")
+                .logContains("This shouldn't show up on second run")
+                .logNotContains("This should only run on restart")
+                .go();
+
+        WorkflowJob p = original.getParent();
+
+        HtmlPage redirect = restartFromStageInUI(original, "restart");
+
+        assertNotNull(redirect);
+        assertEquals(p.getAbsoluteUrl(), redirect.getUrl().toString());
+
+        j.waitUntilNoActivity();
+        WorkflowRun b2 = p.getBuildByNumber(2);
+        assertNotNull(b2);
+        j.assertBuildStatusSuccess(b2);
+        j.assertLogContains("This should only run on restart", b2);
+        j.assertLogNotContains("This shouldn't show up on second run", b2);
+    }
+
+    private HtmlPage restartFromStageInUI(@Nonnull WorkflowRun original, @Nonnull String stageName) throws Exception {
+        RestartDeclarativePipelineAction action = original.getAction(RestartDeclarativePipelineAction.class);
+        assertNotNull(action);
+        assertTrue(action.isRestartEnabled());
+
+        HtmlPage page = j.createWebClient().getPage(original, action.getUrlName());
+        HtmlForm form = page.getFormByName("restart");
+        HtmlSelect select = form.getSelectByName("stageName");
+        select.getOptionByValue(stageName).setSelected(true);
+        return j.submit(form);
+    }
```

Comment:
restart with new Jenkinsfile where stage names has changed?
See my thoughts before.

Reply:
As mentioned above, that isn't actually possible, so we're good. =)